content/2022-02-25-pattern-matching-with-named-fields.md

```diff
@@ -0,0 +1,306 @@
+
+```scala
+  case User(city = city1, city = city2) => a // error city is used twice
+  case User(name1, name = name2) => a // error city is used twice
+```
+
+The same should happen if the `User` had a field with a [deprecated name](https://www.scala-lang.org/api/current/scala/deprecatedName.html).
+
+### Order of name and term
+
+Normally an equal sign assigns the result of the right side to the left side. With the proposed syntax that's not the case. However, for most, if not all, people that's the correct order.
+
+This order seems the requires a single look ahead in the parser.
+
+### Desugaring
+
+> One important principle of Scala’s pattern matching design is that case classes should be abstractable. I.e. we want to be able to represent the abilities of a case class without exposing the case class itself. That also allows code to evolve from a case class to a regular class or a different case class while maintaining the interface of the old case class. [Martin Odersky](https://contributors.scala-lang.org/t/pattern-matching-with-named-fields/1829/52)
+
+* Allow the same feature of case classes:
+  * Provide arbitrary names
+  * Enable [`@deprecatedName`](https://www.scala-lang.org/api/current/scala/deprecatedName.html), which leads to multiple names for a single field. Using both should be an error.
+* Maybe provide additional named fields. For example for our `User`:
+
+  ```scala
+  case User(ageInDays = 3650) => ...
+  ```
+
+  But this could be an awful feature.
+* Maybe leave fields without names
+
+This SIP proposes basically to use [shapeless records](https://github.com/milessabin/shapeless/wiki/Feature-overview:-shapeless-2.0.0#extensible-records) for the specific purpose of pattern matching.
+For incomplete list of alternatives see [alternative desugaring](#alternative-desugaring) below.
+
+Pro:
+
+* with shapeless records we have a good understanding what we are doing, however we need to tweak the encoding a bit
+* with some helper definitions the usage could be quite short
+* reuse possible
+
+Con:
+
+* opens an can of worms, as this encoding could be useful for other requested features like [named tuples](named-tuple), which would require more thought-out design
+* no support for `@deprecatedName`
+* identifiers are represented with string literals
+
+Example for user:
+
+```scala
+object User:
+  def unapply(user: User): Option[
+    (String, Int, String) & 
+    { type Names = ("name", "age", "city") }
+  ] = Some((user.name, user.age, user.city)).asInstanceOf
+```
+
+// TODO: Describe why Names can also be use toplevel.
+
+The reason to pull the names out, instead of keeping them near to their type, is to prevent extractors to accidentally leaking the name.
+
+## Implementation
+
+If a pattern with a name is encountered, the compiler looks up list of provided names and places the trees accordingly.
+
+The list of names either provided by the return type of the unapply method or by the constructor list of the case class.
```

Comment:
This would simplify things. But I have two concerns:
* This would change the signature of all generated `unapply`s. For most cases this should be fine, because the new type is more specific than the old type. I'm unsure about libraries, like slick, which rely on the unapply method. But thats why we have the community build, right?
* The current implementation would retroactively add named pattern for libraries compiled with an older scala version. That's maybe a good thing.

I'll give a try and see how much breaks.